### PR TITLE
Add inverse csr endpoints and methods

### PR DESCRIPTION
**Pull request by ralphiee22**
## Summary

All the relevant pieces have been added to allow a client to push up its data to a server. There is a shared public/private key pair used to instantiate certificates on the client side (using public key), and provide ownership on the server side (using private key). Endpoints for pushing certificate chain and single certificates will only be allowed if there is a `ALLOW_CERTIFICATE_PUSHING` setting on the server. 

## TODO

- [x] Have tests been written for the new code?
- [ ] Has documentation been written/updated?
- [ ] New dependencies (if any) added to requirements file

## Reviewer guidance

*If you PR has a significant size, give the reviewer some helpful remarks*

## Issues addressed

List the issues solved or partly solved by the PR

## Documentation

If the PR has documentation, link the file here (either .rst in your repo or if built on Read The Docs)

### CODE REVIEW

File: morango/syncsession.py [hunk 1]
```diff
@@ -183,13 +183,67 @@ def certificate_signing_request(self, parent_cert, scope_definition_id, scope_pa
...
+            # build up data for csr
+            certificate = Certificate(parent_id=local_parent_cert.id,
+                                      profile=local_parent_cert.profile,
+                                      scope_definition_id=scope_definition_id,
+                                      scope_version=local_parent_cert.scope_version,
+                                      scope_params=json.dumps(scope_params),
+                                      public_key=Key(public_key_string=publickey_response.json()['public_key']),
+                                      )
+            # add ID and signature to the certificate
+            certificate.salt = uuid.uuid4().hex
```
Comment (ralphiee22) [inline morango/syncsession.py]: To avoid collisions, use nonce on server.

File: morango/syncsession.py [hunk 2]
```diff
@@ -183,13 +183,67 @@ def certificate_signing_request(self, parent_cert, scope_definition_id, scope_pa
         csr_cert.save()
         return csr_cert
 
+    def inverse_certificate_signing_request(self, local_parent_cert, scope_definition_id, scope_params, userargs=None, password=None):
+        # check for certs owned by the server
+        cert_resp = self._request(api_urls.CERTIFICATE, params={'parent': local_parent_cert.id})
```
Comment (ralphiee22) [inline morango/syncsession.py]: Remove logic below and move certificate filtering to management command.

File: morango/api/viewsets.py [hunk 3]
```diff
@@ -71,19 +71,36 @@ def create(self, request):
         else:
             return response.Response(serialized_cert.errors, status=status.HTTP_400_BAD_REQUEST)
 
-    @decorators.action(methods=['post'], detail=False)
-    def push_certificate(self, request):
+    @decorators.action(methods=['post'], detail=False, authentication_classes=[], permission_classes=[])
```
Comment (ralphiee22) [inline morango/api/viewsets.py]: Removing permissions to allow anyone to push to a server that has `ALLOW_CERTIFICATE_PUSHING` set to `True`.

File: tests/testapp/tests/test_api.py [hunk 4]
```diff
@@ -836,3 +907,20 @@ def test_id_changes_id_hash_changes(self):
             InstanceIDModel.get_or_create_current_instance()
             m_info = self.client.get(reverse('morangoinfo-detail', kwargs={"pk": 1}), format='json')
         self.assertNotEqual(m_info.data['instance_hash'], old_id_hash)
+
+
+class PublicKeyTestCase(APITestCase):
+
+    def setUp(self):
+        key = Key()
+        self.key = SharedKey.objects.create(public_key=Key(public_key_string=key.get_public_key_string()),
```
Comment (jamalex) [inline tests/testapp/tests/test_api.py]: as above, can just say `public_key=key`

File: tests/testapp/tests/test_connection.py [hunk 5]
```diff
@@ -103,6 +114,24 @@ def test_csr(self):
             self.network_connection.certificate_signing_request(self.root_cert, '', '')
         self.assertTrue(Certificate.objects.filter(id=json.loads(cert_serialized)['id']).exists())
 
+    @mock_patch_decorator
+    def test_push_signed_client_certificate_chain(self):
+        # sanity test to make sure things are working as expected
+        NetworkSyncConnection._request.return_value.json.return_value = {'id': uuid.uuid4().hex, 'public_key': Key().get_public_key_string()}
+        cert = self.network_connection.push_signed_client_certificate_chain(self.root_cert,
+                                                                            self.subset_scope_def.id,
+                                                                            {"mainpartition": self.root_cert.id, "subpartition": "abracadabra"})
```
Comment (jamalex) [inline tests/testapp/tests/test_connection.py]: Wait, so what is this test testing, exactly? The API endpoints are mocked, so it's just testing that the `push_signed_client_certificate_chain` method is able to complete?

File: tests/testapp/tests/test_connection.py [hunk 6]
```diff
@@ -103,6 +114,24 @@ def test_csr(self):
...
+
+    @mock_patch_decorator
+    def test_push_signed_client_certificate_chain_403(self):
+        # access denied when requesting public key endpoint
+        NetworkSyncConnection._request.side_effect = HTTPError(response=mock.Mock(status_code=403))
+        with self.assertRaises(MorangoServerDoesNotAllowNewCertPush):
+            self.network_connection.push_signed_client_certificate_chain(self.root_cert,
+                                                                         self.subset_scope_def.id,
+                                                                         {"mainpartition": self.root_cert.id, "subpartition": "abracadabra"})
+
```
Comment (jamalex) [inline tests/testapp/tests/test_connection.py]: General thought about the tests: You seem to be separately testing either the API endpoints or the code that calls them. Would it be possible to do a few "integration tests" of calling something like `push_signed_client_certificate_chain` and having it hit the endpoints, and make sure they play well together?

File: morango/syncsession.py [hunk 7]
```diff
@@ -183,13 +183,58 @@ def certificate_signing_request(self, parent_cert, scope_definition_id, scope_pa
...
+        certificate.parent.sign_certificate(certificate)
+
+        certificate.save()
+
+        # client sends signed cert to server
+        try:
+            self._push_certificate_chain(certificate)
+        # for any errors, we delete the certificate we attempted to send to the server
+        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError):
+            certificate.delete()
```
[the remark this answers is not in the document]
Comment (ralphiee22) [reply morango/syncsession.py]: the thing with saving the cert is that it allows us to call `.get_descendants()` on it. I suppose before saving we can just call `get_descendants()` on the parent cert and tack on the additional cert we created in the request.

File: tests/testapp/tests/test_connection.py (hunk 5 above)
Comment (ralphiee22) [reply tests/testapp/tests/test_connection.py]: yes was thinking this as a smoke test.

File: morango/syncsession.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamalex) [reply morango/syncsession.py]: Ah, ok -- whichever feels cleaner, then, to you

**Pull request merged**